File: attached_assets/_extracted/Competency-Platform-main/src/server/services/bell-curve.ts
/**
 * Bell-curve / distribution-policy analysis.
 *
 * Ported from the reference analysis tool. Pure & framework-free so it runs on the
 * server (to compute department aggregates) and on the client (to render the
 * per-department comparison without a round-trip), and is unit-testable.
 *
 * Categories are ordered worst → best (index 0..4), matching the policy arrays:
 *   0 غير مرضي · 1 دون المتوقع · 2 حسب المتوقع · 3 فوق المتوقع · 4 استثنائي
 * which align 1:1 with our evaluation rating bands by total score.
 */

export const BELL_CATEGORY_LABELS = [
  "غير مرضي",
  "دون المتوقع",
  "حسب المتوقع",
  "فوق المتوقع",
  "استثنائي",
] as const;

export type AchievementBand = "above" | "achieved" | "below";

export interface PolicySet {
  above: number[];
  achieved: number[];
  below: number[];
}

/** The seeded default policy (also used as a fallback if none is active). */
export const DEFAULT_POLICY: PolicySet = {
  above: [0, 0, 35, 50, 15],
  achieved: [5, 5, 40, 40, 10],
  below: [10, 10, 35, 40, 5],
};

/** Map an evaluation total score (0..100) to a category index (0=worst..4=best). */
export function scoreToBandIndex(score: number): number {
  if (score >= 91) return 4;
  if (score >= 76) return 3;
  if (score >= 61) return 2;
  if (score >= 41) return 1;
  return 0;
}

/** Which policy column applies, based on a department's achievement ratio (1.0 = 100%). */
export function getAchievementCategory(achievement: number): AchievementBand {
  if (achievement > 1.0) return "above";
  if (achievement >= 0.95) return "achieved";
  return "below";
}

export function getPolicyForAchievement(achievement: number, policies: PolicySet): number[] {
  return [...policies[getAchievementCategory(achievement)]];
}

export type ShiftArrow = "up" | "down" | null;

/**
 * Allows the "حسب المتوقع" and "فوق المتوقع" categories to borrow unused quota from
 * the adjacent (better) category, mirroring the reference tool's tolerance rule.
 */
export function calculateShiftedPolicy(
  categories: number[],
  originalPolicy: number[],
): { shifted: number[]; notes: string[]; shiftArrows: ShiftArrow[] } {
  const shifted = [...originalPolicy];
  const notes: string[] = [];
  const shiftArrows: ShiftArrow[] = new Array(5).fill(null);

  for (let i = 2; i < 4; i++) {
    const actual = categories[i] ?? 0;
    const currentPolicy = shifted[i] ?? 0;
    if (actual > currentPolicy) {
      const diff = actual - currentPolicy;
      const nextIdx = i + 1;
      if (nextIdx < 5) {
        const available = shifted[nextIdx] ?? 0;
        const take = Math.min(diff, available);
        if (take > 0) {
          shifted[i] = currentPolicy + take;
          shifted[nextIdx] = available - take;
          shiftArrows[i] = "up";
          shiftArrows[nextIdx] = "down";
          notes.push(
            `فئة "${BELL_CATEGORY_LABELS[i]}" استلفت ${take}% من حصة فئة "${BELL_CATEGORY_LABELS[nextIdx]}"`,
          );
        }
      }
    }
  }
  return { shifted, notes, shiftArrows };
}

/** Departments with fewer than 10 employees are excluded from the mandatory policy. */
export const POLICY_EXCLUSION_THRESHOLD = 10;

export function isDeptCompliant(args: {
  categories: number[];
  employeeCount: number;
  achievement: number;
  policies: PolicySet;
}): boolean {
  if (args.employeeCount < POLICY_EXCLUSION_THRESHOLD) return true;
  const original = getPolicyForAchievement(args.achievement, args.policies);
  const { shifted } = calculateShiftedPolicy(args.categories, original);
  return args.categories.every((val, idx) => val <= (shifted[idx] ?? 0));
}

File: artifacts/api-server/internal/bellcurve/bellcurve.go
// Package bellcurve ports the distribution-policy analysis from the original
// TypeScript service. Pure and dependency-free.
package bellcurve

var CategoryLabels = []string{"غير مرضي", "دون المتوقع", "حسب المتوقع", "فوق المتوقع", "استثنائي"}

// PolicySet holds the three achievement-band distributions.
type PolicySet struct {
        Labels   []string `json:"labels"`
        Above    []int    `json:"above"`
        Achieved []int    `json:"achieved"`
        Below    []int    `json:"below"`
}

// DefaultPolicy is the seeded default (also used as a fallback).
var DefaultPolicy = PolicySet{
        Above:    []int{0, 0, 35, 50, 15},
        Achieved: []int{5, 5, 40, 40, 10},
        Below:    []int{10, 10, 35, 40, 5},
}

// ScoreToBandIndex maps a 0..100 score to a category index (0=worst..4=best).
func ScoreToBandIndex(score int) int {
        switch {
        case score >= 91:
                return 4
        case score >= 76:
                return 3
        case score >= 61:
                return 2
        case score >= 41:
                return 1
        default:
                return 0
        }
}

// AchievementCategory returns which policy column applies for a ratio (1.0=100%).
func AchievementCategory(achievement float64) string {
        switch {
        case achievement > 1.0:
                return "above"
        case achievement >= 0.95:
                return "achieved"
        default:
                return "below"
        }
}

func PolicyForAchievement(achievement float64, p PolicySet) []int {
        switch AchievementCategory(achievement) {
        case "above":
                return append([]int(nil), p.Above...)
        case "achieved":
                return append([]int(nil), p.Achieved...)
        default:
                return append([]int(nil), p.Below...)
        }
}

// ShiftResult is returned by CalculateShiftedPolicy.
type ShiftResult struct {
        Shifted     []int    `json:"shifted"`
        Notes       []string `json:"notes"`
        ShiftArrows []string `json:"shiftArrows"` // "up" | "down" | ""
}

// CalculateShiftedPolicy lets the middle categories borrow unused quota from the
// adjacent better category (tolerance rule from the reference tool).
func CalculateShiftedPolicy(categories, originalPolicy []int) ShiftResult {
        shifted := append([]int(nil), originalPolicy...)
        for len(shifted) < 5 {
                shifted = append(shifted, 0)
        }
        notes := []string{}
        arrows := make([]string, 5)

        at := func(s []int, i int) int {
                if i < len(s) {
                        return s[i]
                }
                return 0
        }

        for i := 2; i < 4; i++ {
                actual := at(categories, i)
                current := shifted[i]
                if actual > current {
                        diff := actual - current
                        next := i + 1
                        if next < 5 {
                                available := shifted[next]
                                take := diff
                                if available < take {
                                        take = available
                                }
                                if take > 0 {
                                        shifted[i] = current + take
                                        shifted[next] = available - take
                                        arrows[i] = "up"
                                        arrows[next] = "down"
                                        notes = append(notes, "فئة \""+CategoryLabels[i]+"\" استلفت من حصة فئة \""+CategoryLabels[next]+"\"")
                                }
                        }
                }
        }
        return ShiftResult{Shifted: shifted, Notes: notes, ShiftArrows: arrows}
}

// ExclusionThreshold: departments with fewer employees are exempt.
const ExclusionThreshold = 10

// IsDeptCompliant reports whether a department's distribution respects policy.
func IsDeptCompliant(categories []int, employeeCount int, achievement float64, p PolicySet) bool {
        if employeeCount < ExclusionThreshold {
                return true
        }
        original := PolicyForAchievement(achievement, p)
        res := CalculateShiftedPolicy(categories, original)
        for i, val := range categories {
                limit := 0
                if i < len(res.Shifted) {
                        limit = res.Shifted[i]
                }
                if val > limit {
                        return false
                }
        }
        return true
}
